File: src/test/scala/com/exasol/cloudetl/scriptclasses/StorageTest.scala
package com.exasol.cloudetl.scriptclasses

import java.nio.file.Path
import java.nio.file.Paths

import com.exasol.ExaIterator
import com.exasol.cloudetl.storage.StorageProperties

import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.when
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.mockito.MockitoSugar

/**
 * A trait with helper methods and default values for storage test
 * classes.
 */
trait StorageTest extends AnyFunSuite with MockitoSugar {

  private[scriptclasses] val testResourceDir: String =
    normalize(Paths.get(getClass.getResource("/data").toURI))

  private[scriptclasses] val testResourceParquetPath: String =
    s"$testResourceDir/import/parquet/sales_pos*.parquet"

  final def normalize(path: Path): String =
    path.toUri.toString.replaceAll("/$", "").replaceAll("///", "/")

  final def mockExasolIterator(params: Map[String, String]): ExaIterator = {
    val storageProperties = StorageProperties(params)
    val bucketPath = storageProperties.getStoragePath()

    val mockedIterator = mock[ExaIterator]
    when(mockedIterator.getString(0)).thenReturn(bucketPath)
    when(mockedIterator.getString(1)).thenReturn(storageProperties.mkString())
    mockedIterator
  }

  final def anyObjects(): Array[Object] =
    any(classOf[Array[Object]])

}

File: src/test/java/com/exasol/cloudetl/scriptclasses/StorageTest.java
package com.exasol.cloudetl.scriptclasses;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

import com.exasol.ExaIterator;
import com.exasol.cloudetl.storage.StorageProperties;

abstract class StorageTest {
    protected final String testResourceDir = normalize(resourcePath("/data"));
    protected final String testResourceParquetPath = this.testResourceDir + "/import/parquet/sales_pos*.parquet";

    protected String normalize(final Path path) {
        return path.toUri().toString().replaceAll("/$", "").replaceAll("///", "/");
    }

    protected ExaIterator mockExasolIterator(final Map<String, String> params) {
        final StorageProperties storageProperties = new StorageProperties(params);
        final String bucketPath = storageProperties.getStoragePath();
        final ExaIterator mockedIterator = mock(ExaIterator.class);
        try {
            when(mockedIterator.getString(0)).thenReturn(bucketPath);
            when(mockedIterator.getString(1)).thenReturn(storageProperties.mkString());
        } catch (final com.exasol.ExaIterationException | com.exasol.ExaDataTypeException exception) {
            throw new IllegalStateException(exception);
        }
        return mockedIterator;
    }

    protected Object[] anyObjects() {
        return any(Object[].class);
    }

    private static Path resourcePath(final String resource) {
        try {
            return Paths.get(StorageTest.class.getResource(resource).toURI());
        } catch (final URISyntaxException exception) {
            throw new IllegalStateException(exception);
        }
    }
}
